### PR TITLE
Adopt strict batch shape semantics for distributions

### PR DESCRIPTION
**Pull request by fritzo**
See [Design Doc](https://docs.google.com/document/d/1EpAnAIotJj24-yS_8h_5KEiYWYop7zHcAErSeO7XAZc/edit) | Blocking #780

This PR adds checks for stricter use of batch dimensions in Pyro. This is only recently possible since it relies on PyTorch support for scalars.

## Why?

Pyro 0.2 will be able to automatically introduce batch dimensions for parallelizing `enum_discrete` and `num_particles`. To give Pyro space to create new batch dimensions, we need to restrict how dimensions are introduced by user code.

## How?

The new requirements are:
- users must declare non-batch dimensions using `.reshape(extra_event_dims=n)`
- batch dimensions will only be allowed inside `iarange` (and maybe `irange`?)
- to use auto-parallelization, users must request dims via `SVI(..., max_iarange_nesting=n)`

These requirements are checked by the new helper `check_site_shape` which uses the new `.size` field in `cond_indep_stack` frames.

## Tasks

- [x] check sites in `Trace_ELBO`
- [x] check sites in `TraceGraph_ELBO`
- [x] update tests/infer/test_valid_models.py
- [x] update other tests
- [x] update examples
- [x] update tutorials
- [ ] document new batching semantics

### CODE REVIEW

**Comment by fritzo** [comment]
@neerajprad I've merged dev with your #824 into this PR. Lots of tests still fail. Feel free to push commits to this branch. I'll chat you if I start working on it again (at earliest Wednesday).

File: pyro/util.py [hunk 1]
```diff
@@ -354,6 +349,35 @@ def check_model_guide_match(model_trace, guide_trace):
...
+            raise ValueError('\n  '.join([
+                'at site "{}", invalid log_prob shape'.format(site["name"]),
+                'Expected shape compatible with {}, actual {}'.format(expected_shape, actual_shape),
+                'Try one of the following fixes:',
+                '- enclose the batched tensor in a with iarange(...): context',
+                '- .reshape(extra_event_dims=...) the distribution being sampled',
+                '- .permute() data dimensions']))
+
+    # Check parallel dimensions on the left of max_iarange_nesting.
+    # TODO
```
Comment (neerajprad) [inline pyro/util.py]: @fritzo : Do you want to add some more checks in this PR?

File: tests/infer/test_inference.py [hunk 2]
```diff
@@ -371,8 +374,8 @@ def model():
             mu_latent = pyro.sample("mu_latent",
                                     dist.Normal(self.mu0, torch.pow(self.tau0, -0.5)))
             sigma = torch.pow(self.tau, -0.5)
-            pyro.observe("obs0", dist.LogNormal(mu_latent, sigma), obs=self.data[0])
-            pyro.observe("obs1", dist.LogNormal(mu_latent, sigma), obs=self.data[1])
+            pyro.observe("obs0", dist.LogNormal(mu_latent, sigma), obs=self.data[0].squeeze())
```
Comment (neerajprad) [inline tests/infer/test_inference.py]: This test fails to meet the threshold if we use a vectorized "obs" inside an `iarange`. 

@martinjankowiak, @frtizo - Is that expected? Is there a reason why the two `obs` are observed in separate statements?

File: pyro/distributions/testing/rejection_gamma.py [hunk 3]
```diff
@@ -18,7 +18,7 @@ def __init__(self, alpha):
         if alpha.data.min() < 1:
             raise NotImplementedError('alpha < 1 is not supported')
         self.alpha = alpha
-        self._standard_gamma = Gamma(alpha, alpha.new([1]).expand_as(alpha))
+        self._standard_gamma = Gamma(alpha, torch.empty_like(alpha).fill_(1).expand_as(alpha))
```
Comment (neerajprad) [inline pyro/distributions/testing/rejection_gamma.py]: @fritzo - Had to make a small change to support scalars. Is there a more concise way to do this?

**Comment by neerajprad** [comment]
The times for integration tests seem to have substantially increased with either the checks or the iarange/reshaping operations.

File: tests/infer/test_inference.py (hunk 2 above)
Comment (neerajprad) [reply tests/infer/test_inference.py]: Just checked with @martinjankowiak that this works as expected inside `iarange`, so I am not sure why I was seeing a difference. Will make the change and update.

File: pyro/util.py (hunk 1 above)
Comment (fritzo) [reply pyro/util.py]: No, they should be added in later PRs.

File: pyro/distributions/testing/rejection_gamma.py (hunk 3 above)
Comment (fritzo) [reply pyro/distributions/testing/rejection_gamma.py]: How about `alpha.new([1]).squeeze().expand_as(alpha)`?

File: tests/infer/test_inference.py (hunk 2 above)
Comment (fritzo) [reply tests/infer/test_inference.py]: @martinjankowiak can we just delete this flaky expensive test?

**Comment by fritzo** [comment]
@neeraj, is this ready to merge?

File: tests/infer/test_elbo_mapdata.py [hunk 4]
```diff
@@ -18,7 +18,7 @@
 @pytest.mark.stage("integration", "integration_batch_1")
 @pytest.mark.init(rng_seed=161)
 @pytest.mark.parametrize("batch_size", [3, 5, 7, 8, None])
-@pytest.mark.parametrize("map_type", ["tensor", "list"])
+@pytest.mark.parametrize("map_type", ["iarange", "irange", "range"])
```
Comment (fritzo) [inline tests/infer/test_elbo_mapdata.py]: @neerajprad We're actually running more tests in integration_batch_1, so that might be part of the slowdown.

**Comment by neerajprad** [comment]
@martinjankowiak - could you take a look at this PR? It will be nice to get this merged soon before we accumulate merge conflicts.

**Comment by martinjankowiak** [comment]
yeah i'll take a look now

File: tests/infer/test_inference.py (hunk 2 above)
Comment (martinjankowiak) [reply tests/infer/test_inference.py]: yes we can. although maybe keep the transformed_distribution bit?

File: examples/vae.py [hunk 5]
```diff
@@ -121,9 +123,9 @@ def model_sample(self, batch_size=1):
         # sample the handwriting style from the constant prior distribution
         prior_mu = Variable(torch.zeros([batch_size, self.z_dim]))
         prior_sigma = Variable(torch.ones([batch_size, self.z_dim]))
-        zs = pyro.sample("z", dist.Normal(prior_mu, prior_sigma))
+        zs = pyro.sample("z", dist.Normal(prior_mu, prior_sigma).reshape(extra_event_dims=2))
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply examples/vae.py]: I think that's an oversight on my part. We don't need any reshaping here. Will remove.

File: examples/vae.py [hunk 6]
```diff
@@ -123,9 +123,9 @@ def model_sample(self, batch_size=1):
         # sample the handwriting style from the constant prior distribution
         prior_mu = Variable(torch.zeros([batch_size, self.z_dim]))
         prior_sigma = Variable(torch.ones([batch_size, self.z_dim]))
-        zs = pyro.sample("z", dist.Normal(prior_mu, prior_sigma).reshape(extra_event_dims=2))
+        zs = pyro.sample("z", dist.Normal(prior_mu, prior_sigma))
```
Comment (fritzo) [inline examples/vae.py]: Why don't you need to `.reshape(extra_event_dims=1)` here?

Comment (neerajprad) [reply examples/vae.py]: From my understanding, `model_sample` is not used for inference but for generating samples separately using the trained `decoder`, for plotting, etc.

**Pull request merged**